### ISSUE TITLE
bf1942 Player Keyhash Issue

### ISSUE THREAD

**Issue by RestNPizza** [gamedig/node-gamedig#337]
I've noticed an issue with the bf1942 game type where the player keyhash that returns for a player is cut off if it starts with a number. Keyhashes that start with non-numerical characters are displayed in their entirety but keyhashes that start with a number only show the keyhash up to the first non-numerical character. For example, if my keyhash was "abc3433233123" it would show the complete keyhash but if my keyhash was "34abc34332331" it would show my keyhash as "34". Any chance this could be fixed?

**Comment by cetteup** [comment]
Do you have an example? None of the Battlefield 1942 servers that I tried returning anything that looks like a key hash for players.

Try running a query with `--debug` and post the output.

**Comment by cetteup** [comment]
The issue is likely related to this section in the GameSpy 1 protocol.

https://github.com/gamedig/node-gamedig/blob/896a101b7a10ec4ef6e59f4ac5c6c56ce73493b8/protocols/gamespy1.js#L79-L82

If the value starts with a non-numeric character, `parseInt` returns `NaN` and the value is added as a string. However, if the value instead starts with a numeric character, `parseInt` returns any numbers up to the first non-numeric character.
```js
const parsed = parseInt('123abc')
console.log(parsed) // will log 123
```

The easiest fix would most likely be to replace `!isNaN(parseInt(value))` with `!isNaN(Number(value))`. The later will return `NaN` for a string that contains any non-numeric characters.

```js
const converted = Number('123abc')
console.log(converted) // will log NaN
```

@mmorrisontx This may be an issue in a few placed. `parseInt` is a bit odd in this regard. Are you aware of any scenario where the described behavior would be desired?

**Comment by dgwood12** [comment]
> Do you have an example? None of the Battlefield 1942 servers I tried returned anything that looks like a key hash for players.
> 
> Try running a query with `--debug` and post the output.

This server shows keyhashes: 51.81.48.224:23000

Here's the debug output:
[output.txt](https://github.com/gamedig/node-gamedig/files/11174483/output.txt)

**Comment by cetteup** [comment]
That confirms what I described above, as the cut off keys have been parsed to integers.

```json
{
  "name": "W",
  "raw": {
    "deaths": 0,
    "keyhash": 8515,
    "kills": 0,
    "ping": 128,
    "score": 0,
    "teamId": 0,
    "team": "axis"
  }
}
```

Will need to wait for feedback from @mmorrisontx, as I am unsure whether the proposed change would break the implementation for other games.

**Comment by RestNPizza** [comment]
> That confirms what I described above, as the cut off keys have been parsed to integers.
> 
> ```json
> {
>   "name": "W",
>   "raw": {
>     "deaths": 0,
>     "keyhash": 8515,
>     "kills": 0,
>     "ping": 128,
>     "score": 0,
>     "teamId": 0,
>     "team": "axis"
>   }
> }
> ```
> 
> Will need to wait for feedback from @mmorrisontx, as I am unsure whether the proposed change would break the implementation for other games.

Not trying to be pushy but just wondering if this was looked at/discussed yet. Thanks for all your hard work!

**Comment by RestNPizza** [comment]
Has this project been abandoned? :( Seems like there's been quite a bit of time that has passed since things have been updated last.

**Comment by cetteup** [comment]
Any idea whether replacing `parseInt` with `Number` is safe here? @CosminPerRam

**Comment by CosminPerRam** [comment]
Yep, just that `Number` also parses floats, but since working on [rust-GameDig](github.com/gamedig/rust-gamedig), i haven't seen a GameSpy key that is a float, so a `Math.floor` is also needed here, will make a PR to fix this soon.

**Comment by RestNPizza** [comment]
> Yep, just that `Number` also parses floats, but since working on [rust-GameDig](github.com/gamedig/rust-gamedig), i haven't seen a GameSpy key that its value to be float, and a `Math.floor` might be needed here eventually for the teams part (?), will make a PR to fix this soon.

Thank you so much!! I look forward to the fix!

**Comment by cetteup** [comment]
I assume the cleanest way to implement this would be
```javascript
if (key !== 'name' && !isNaN(Number(value).toFixed())) { 
    value = Number(value).toFixed(); 
}
playersById[id][key] = value;
 ```

`Number(value).toFixed()` returns `NaN` if `value` is not a valid number, so it works without any special handling.

**Comment by CosminPerRam** [comment]
The problem actually comes from here:
https://github.com/gamedig/node-gamedig/blob/4365f4ae0aee38f6d9baed3de5b07cd05e06c63a/protocols/gamespy1.js#L24-L26
When normalizing fields, it parses the keyhash, sees that its not a NaN, and goes ahead.

**Comment by CosminPerRam** [comment]
This is now fixed by #342.
Feel free to reopen if issues/questions appear.
